repair-missing/script.whizzml

```diff
@@ -0,0 +1,192 @@
+;; This is a hack:  Field IDs are not conserved at batch prediction time, so
+;; we need to make sure that all of the fields have assigned IDs exactly as
+;; batch prediction will assign them (which essentially amounts to putting
+;; generated fields at the end).  To do this, we make a dummy model, do a dummy
+;; batch prediction, then remove the dummy column.
+(define (flatten-dataset dataset-id)
+  (let (fds ((fetch dataset-id) "fields"))
+    (if (some (lambda (fid) (= ((fds fid) "optype") "datetime")) (keys fds))
+      (let (_ (log-info "Generated fields detected.  Flattening dataset...")
+            mod-id (create-model dataset-id)
+            bp (create-batchprediction {"dataset" dataset-id
+                                        "all_fields" true
+                                        "output_dataset" true
+                                        "model" (wait mod-id)
+                                        "prediction_name" "__bmldummy"})
+            pred-ds ((fetch (wait bp)) "output_dataset_resource")
+            new_fields ((fetch (wait pred-ds)) "fields")
+            pred_fid (id-from-fields new_fields "__bmldummy")
+            final-ds (wait (create-dataset {"origin_dataset" pred-ds
+                                            "excluded_fields" [pred_fid]})))
+        (delete* [mod-id pred-ds bp])
+        final-ds)
+      dataset-id)))
+
+;; Helper function to get the ID of a field given its name
+(define (id-from-fields fields name)
+  (loop (ids (keys fields) vs (values fields))
+    (cond (empty? ids) false
+          (= name ((head vs) "name" false)) (head ids)
+          (recur (tail ids) (tail vs)))))
+
```

Comment:
Not at all a newbie question!  I copy-pasted this from a time before we built in `filter`, which would indeed be better here.  Though in this case (selecting one item), I prefer `some`.  I'll change it.